### PR TITLE
Use lazy inputs in building

### PR DESCRIPTION
**Pull request by mforets**
In master:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 4-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(1.162 s)

          ("BLDC01-BDS01", "dense") => TrialEstimate(1.160 s)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(1.106 s)
          ("BLDC01-BDS01", "discrete") => TrialEstimate(1.125 s)
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 4-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(1.218 s)
          ("BLDC01-BDS01", "dense") => TrialEstimate(1.204 s)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(1.181 s)
          ("BLDC01-BDS01", "discrete") => TrialEstimate(1.154 s)
```

This branch, that uses lazy inputs to improve the accuracy hence the time step can be increased:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 4-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(218.159 ms)

          ("BLDC01-BDS01", "dense") => TrialEstimate(471.524 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(209.145 ms)
          ("BLDC01-BDS01", "discrete") => TrialEstimate(469.710 ms)
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 4-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(227.663 ms)
          ("BLDC01-BDS01", "dense") => TrialEstimate(484.326 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(216.985 ms)
          ("BLDC01-BDS01", "discrete") => TrialEstimate(495.111 ms)
```

### CODE REVIEW

**Comment by mforets** [comment]
`:set_type=>Interval` doesn't work because the partition is high-dimensional:

```julia
julia> include("models/ARCH/AFF/Building/building_benchmark.jl")
Computing successors 100%|██████████████████████████████| Time: 0:00:02
ERROR: LoadError: LoadError: AssertionError: a linear map of size (48, 48) cannot be applied to a set of dimension 1
```

But we can use hyperrectangle for the decomposition of initial states and interval for the iteration, like:

```julia
𝑂_dense = Options(:δ=>0.004, :vars=>[25], :set_type_init=>Hyperrectangle, :set_type_iter=>Interval, :partition=>[1:48], :lazy_inputs_interval=>-1)
```

File: models/ARCH/AFF/Building/building_BFFPSV18.jl [hunk 1]
```diff
@@ -8,7 +8,7 @@ include("building.jl")
 𝑂_BLDF01 = Options(:T=>time_horizon, :mode=>"check", :property => pBDS01)
 
 # algorithm-specific options
-𝑂_dense = Options(:δ=>0.004, :vars=>[25], :set_type=>Interval, :set_type=>Hyperrectangle, :partition=>[1:48], :lazy_inputs_interval=>-1)
```
Comment (schillic) [inline models/ARCH/AFF/Building/building_BFFPSV18.jl]: This was the wrong `set_type` you removed :smile:

Comment (schillic) [reply models/ARCH/AFF/Building/building_BFFPSV18.jl]: Actually, you can (and should) remove `:set_type` altogether for the given partition. It does not have any effect: one-block partitions are not decomposed, and `"check"` mode does not overapproximate.

Comment (schillic) [reply models/ARCH/AFF/Building/building_BFFPSV18.jl]: > There is something I do not understand: `set_type=>Hyperrectangle` should be _equivalent_ to 1D `Interval`s. How come there is a difference?

That is also explained by this observation.

I find this confusing and created https://github.com/JuliaReach/Reachability.jl/issues/490.

**Comment by mforets** [comment]
Note: i used this change in `Reachability.jl#master`:

```julia
$ git diff
diff --git a/src/ReachSets/ContinuousPost/BFFPSV18/partitions.jl b/src/ReachSets/ContinuousPost/BFFPSV18/partitions.jl
index 3f0b7ad..599a83c 100644
--- a/src/ReachSets/ContinuousPost/BFFPSV18/partitions.jl
+++ b/src/ReachSets/ContinuousPost/BFFPSV18/partitions.jl
@@ -29,7 +29,8 @@ function inout_map_property(𝑃::PROPERTY,
                             n::Int
                            )::PROPERTY where {PROPERTY<:Property}
     proj = projection_map(partition, blocks)
-    if length(proj) == n
+
+    if isa(𝑃.safe, HalfSpace) || length(proj) == n
         # no change in the dimension, return the original property
         return 𝑃
     else
```

**Comment by mforets** [comment]
After the fixes, the times [in the first comment](https://github.com/JuliaReach/ReachabilityBenchmarks/pull/76#issue-255166556) are no longer valid -- it does not help to use lazy input interval option because for the dense time case, (for example `sol_BLDC01_dense`) i get that the property is invalidated if `:δ=>0.001`. Actually, the transition seems to be between `:δ=>0.0009` (valid) and `:δ=>0.001` (invalid). So i would keep it as is.

The times are:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 4-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(1.238 s)
          ("BLDC01-BDS01", "dense") => TrialEstimate(1.183 s)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(1.069 s)
          ("BLDC01-BDS01", "discrete") => TrialEstimate(1.032 s)
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 4-element BenchmarkTools.BenchmarkGroup:
          tags: []

          ("BLDF01-BDS01", "dense") => TrialEstimate(1.240 s)
          ("BLDC01-BDS01", "dense") => TrialEstimate(1.219 s)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(1.122 s)
          ("BLDC01-BDS01", "discrete") => TrialEstimate(1.055 s)
```

File: models/ARCH/AFF/Building/building.jl [hunk 2]
```diff
@@ -31,22 +31,21 @@ S = ConstrainedLinearControlContinuousSystem(A, B, X, Uin)
...
 radius_X0 = [fill(0.000025, 10); fill(0.0, 14); 0.0001; fill(0.0, 23)]
 X0 = Hyperrectangle(center_X0, radius_X0)
-
-δ_max = 0.0009
-time_horizon = 20.0
 build_TV = InitialValueProblem(S, X0)
 
 # specifications
-pBDS01 = LinearConstraintProperty(sparsevec([25], [1.0], 48), 0.0051) # x25 <= 0.0051
+pBDS01 = SafeStatesProperty(LinearConstraint(sparsevec([25], [1.0], 48), 0.0051)) # x25 <= 0.0051
```
Comment (schillic) [inline models/ARCH/AFF/Building/building.jl]:
```suggestion
pBDS01 = SafeStatesProperty(LinearConstraint(sparsevec([25], [1.0], n), 0.0051)) # x25 <= 0.0051
```

File: models/ARCH/AFF/Building/building_BFFPSV18.jl [hunk 3]
```diff
@@ -8,8 +8,8 @@ include("building.jl")
 𝑂_BLDF01 = Options(:T=>time_horizon, :mode=>"check", :property => pBDS01)
 
 # algorithm-specific options
-𝑂_dense = Options(:δ=>δ_max, :vars=>[25], :set_type=>Interval)
-𝑂_discrete = Options(:δ=>δ_max, :vars=>[25], :set_type=>Interval, :approx_model=>"nobloating")
+𝑂_dense = Options(:δ=>0.0009, :vars=>[25], :lazy_inputs_interval=>-1) # is block options needed at all ? => default is Interval?
```
Comment (schillic) [inline models/ARCH/AFF/Building/building_BFFPSV18.jl]:
```suggestion
𝑂_dense = Options(:δ=>0.0009, :vars=>[25], :lazy_inputs_interval=>-1)
```

**Comment by schillic** [comment]
Do you know what caused the change? Was it a fix or did we just change some defaults? In the latter case, maybe we can get the old (fast) behavior back?

**Comment by mforets** [comment]
It was (more than one) fix. I was able to increase delta only because the property checking was not working properly, after we began using subset check instead of support function check. Note that it would have been detected immediately if we had [this dimensional assertion](https://github.com/JuliaReach/LazySets.jl/pull/1208) already in place.

**Comment by schillic** [comment]
I tried with a one-block partition and a three-block partition. With the one-block partition I could still increase the time step. With the three-block partition, I needed to use lazy decomposition. The fastest settings are benchmark 5. (one-block) and 8. (three-block) below. I propose to use 8. because it is slightly faster and makes proper use of our algorithm.

1. This current branch:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(1.387 s)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(1.223 s)
```

2. `:partition => [1:48]`:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(1.098 s)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(960.763 ms)
```

3. `:partition => [1:48], :lazy_inputs_interval=>0`:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(3.907 s)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(1.813 s)
```

4. `:partition => [1:48], :lazy_inputs_interval=>0, :δ=>0.004`:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(909.582 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(489.887 ms)
```

5. `:partition => [1:48], :δ=>0.004`:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(280.680 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(272.196 ms)
```

6. `:partition => [1:48], :δ=>0.004, :block_options_init => LazySets.LinearMap`:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(911.222 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(489.979 ms)
```

7. `:partition => [1:24, [25], 26:48], :δ=>0.0001`:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(635.228 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(555.378 ms)
```

8. `:partition => [1:24, [25], 26:48], :δ=>0.003, :block_options_init => LazySets.LinearMap`:

```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 2-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(262.804 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(222.617 ms)
```

**Comment by schillic** [comment]
new results:
```julia
minimum: 1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "Build" => 4-element BenchmarkTools.BenchmarkGroup:
          tags: []
          ("BLDF01-BDS01", "dense") => TrialEstimate(270.743 ms)
          ("BLDC01-BDS01", "dense") => TrialEstimate(179.074 ms)
          ("BLDF01-BDS01", "discrete") => TrialEstimate(154.010 ms)
          ("BLDC01-BDS01", "discrete") => TrialEstimate(150.256 ms)
```

**Pull request merged**